format/caff/caff.go

```diff
@@ -0,0 +1,210 @@
+	isObfuscated   bool
+	compressOption uint8
+}
+
+func decodeVersion(d *decode.D) {
+	d.FieldU8("major")
+	d.FieldU8("minor")
+	d.FieldU8("patch")
+}
+
+func decodeCAFF(d *decode.D) any {
+	var ci format.CAFF_In
+	d.ArgAs(&ci)
+
+	var obfsKey int64
+
+	obfsU8 := func(d *decode.D) uint64 { return d.U8() ^ uint64(obfsKey&0xff) }
+	obfsU32 := func(d *decode.D) uint64 { return d.U32() ^ uint64(obfsKey&0xffff_ffff) }
+	obfsU64 := func(d *decode.D) uint64 { return d.U64() ^ uint64(obfsKey<<32|obfsKey) }
+	obfsBool := func(d *decode.D) bool { return obfsU8(d) != 0 }
+
+	// "Big Endian Base 128" - LEB128's strange sibling
+	obfsBEB128 := func(d *decode.D) (v uint64) {
+		for {
+			x := obfsU8(d)
+			v <<= 7
+			v |= (x & 0x7f)
+			if (x >> 7) == 0 {
+				return
+			}
+		}
+	}
+
+	obfsVarStr := func(d *decode.D) string {
+		length := obfsBEB128(d)
+		if length == 0 {
+			return ""
+		}
+
+		raw := d.BytesLen(int(length))
+		for i := uint64(0); i < length; i++ {
+			raw[i] ^= byte(obfsKey)
+		}
+		return string(raw)
+	}
+
+	d.FieldUTF8("archive_id", 4, d.StrAssert("CAFF"))
+	d.FieldStruct("archive_version", decodeVersion)
+	d.FieldUTF8("format_id", 4)
+	d.FieldStruct("format_version", decodeVersion)
+	obfsKey = d.FieldS32("obfuscate_key")
+	d.SeekRel(8 * 8)
+
+	d.FieldStruct("preview_image", func(d *decode.D) {
+		d.FieldU8("image_format", imageFormatNames)
+		d.FieldU8("color_type", colorTypeNames)
+		d.SeekRel(2 * 8)
+		d.FieldU16("width")
+		d.FieldU16("height")
+		d.FieldU64("start_pos")
+		d.FieldU32("file_size")
+	})
+	d.SeekRel(8 * 8)
+
+	fileInfoListSize := d.FieldUintFn("file_info_map_size", obfsU32)
+	fileInfoList := make([]fileInfoListEntry, int(fileInfoListSize))
+
+	d.FieldArray("file_info_list", func(d *decode.D) {
+		for i := uint64(0); i < fileInfoListSize; i++ {
+			d.FieldStruct("file_info", func(d *decode.D) {
+				var entry fileInfoListEntry
+
+				entry.filePath = d.FieldStrFn("file_path", obfsVarStr)
+				d.FieldStrFn("tag", obfsVarStr)
+				entry.startPos = int64(d.FieldUintFn("start_pos", obfsU64))
+				entry.fileSize = int(d.FieldUintFn("file_size", obfsU32))
+				entry.isObfuscated = d.FieldBoolFn("is_obfuscated", obfsBool)
+				entry.compressOption = uint8(d.FieldUintFn("compress_option", obfsU8, compressOptionNames))
+				d.SeekRel(8 * 8)
+
+				fileInfoList[int(i)] = entry
+			})
+		}
+	})
+
+	d.FieldArray("files", func(d *decode.D) {
+		for _, entry := range fileInfoList {
+			d.FieldStruct("file", func(d *decode.D) {
+				d.SeekAbs(entry.startPos * 8)
+				d.FieldValueStr("file_path", entry.filePath)
+				d.FieldValueUint("file_size", uint64(entry.fileSize))
+				d.FieldValueBool("is_obfuscated", entry.isObfuscated)
+				d.FieldValueUint("compress_option", uint64(entry.compressOption), compressOptionNames)
+
+				rawBytes := d.BytesLen(entry.fileSize)
+				if entry.isObfuscated {
+					for i, v := range rawBytes {
+						rawBytes[i] = v ^ uint8(obfsKey)
+					}
+				}
+
+				br := bitio.NewBitReader(rawBytes, -1)
```

Comment:
Sorry for confusing APIs for doing this atm. Hopefully in the future there could be some helpers for dealing with compressed/massaged data